### PR TITLE
Feature/inline input

### PR DESCRIPTION
**Pull request by vinhkq**
Dear Mr Hosoda, i resolved issue 106, detail:
_Make inline input.
_Submit & create todo when enter or leave textbox like gmail todo.
_Add new record to list for refresh todo list.
Please, review & merge.
It was just recently completed, so feedback to me if there are problem.
Thank you.

### CODE REVIEW

File: src/common/api.js [hunk 1]
```diff
@@ -12,9 +12,9 @@ export function createRequestTypes(base) {
   return res;
 }
```
Comment (kenji4569) [inline src/common/api.js]: @vinhkq `option` -> `options`

File: src/common/api.js [hunk 2]
```diff
@@ -27,7 +27,11 @@ export function callApi(path) {
 export function* fetchApi(requestTypes, path, action) {
   try {
     path = format(path, action)
-    const data = yield call(callApi, path);
+    let option = {}
```
Comment (kenji4569) [inline src/common/api.js]: @vinhkq  I don't understand why you insert `action.data.title !== ""`, a such too specific condition, here.

File: src/common/api.js (hunk 1 above)
Comment (kenji4569) [inline src/common/api.js:14]: `option` -> `options`

File: src/pages/todo/sagas.js [hunk 3]
```diff
@@ -3,13 +3,18 @@ import { fork } from 'redux-saga/effects'
...
+  UPDATE_TODO,
+  DELETE_TODO,
+  MOVE_TODO
 } from './actions';
 
 export default function* todoSagas() {
   yield [
     fork(watchApi, LIST_TODOS, '/v1/todos'),
-    fork(watchApi, VIEW_TODO, '/v1/todos/${id}')
+    fork(watchApi, VIEW_TODO, '/v1/todos/${id}'),
```
Comment (kenji4569) [inline src/pages/todo/sagas.js:16]: I rethink that
`fork(watchApi, CREATE_TODO, '/v1/todos'),`
should be `fork(watchApi, CREATE_TODO, '/v1/todos', 'post'),`

and then in common/api.js `export function* watchApi(requestTypes, path)` should be 

```
export function* watchApi(requestTypes, path, method='get') {
  yield* takeLatest(requestTypes.REQUEST, fetchApi, requestTypes, path, method)
}
```

and `export function* fetchApi(requestTypes, path, action)` should be 

```
export function* fetchApi(requestTypes, path, action, method) {
  try {
    path = format(path, action)
    const options = getFetchOptions(method, action);
    const data = yield call(callApi, path, options);
    yield put({type: requestTypes.SUCCESS, data});
  } catch (error) {
    yield put({type: requestTypes.FAILURE, error});
  }
}
```

and `function optionFilter(data)` should be 

```
function getFetchOptions(method, action) {
  let headers = Object.assign({'Accept': 'application/json'}, action.headers)
  let data = action.data
  if (!action.raw) {
    headers = Object.assign(headers, {'Content-Type': 'application/json'})
    data = action.data ? JSON.stringify(action.data) : null
  }
  return {
    method: method,
    headers: headers,
    body: data
  }
}
```

File: src/pages/todo/components/TodoList.js [hunk 4]
```diff
@@ -13,18 +14,76 @@ import { LIST_TODOS } from '../actions';
   };
 })
 class TodoList extends React.Component {
+  constructor(props) {
+    super(props);
+    this.state = {
```
Comment (kenji4569) [inline src/pages/todo/components/TodoList.js:20]: rename `hidden` to `isEditing` (and invert true/false).

File: src/pages/todo/components/TodoList.js [hunk 5]
```diff
@@ -13,18 +14,76 @@ import { LIST_TODOS } from '../actions';
...
+      hidden: true,
+      form: {
```
Comment (kenji4569) [inline src/pages/todo/components/TodoList.js:22]: Remove `method: 'POST'`, this is no longer required.

File: src/pages/todo/components/TodoListItem.js [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+
+  render() {
+    const { todo } = this.props;
+    const { hidden } = this.state;
+    const style = {
+      label: {display: !hidden ? 'none' : 'block'},
+      textbox: {display: hidden ? 'none' : 'block'},
+    }
+    return (
+      <div>
```
Comment (kenji4569) [inline src/pages/todo/components/TodoListItem.js:29]: Is this implementation completed? How to submit the input value?

File: src/pages/todo/components/TodoList.js [hunk 7]
```diff
@@ -13,18 +14,76 @@ import { LIST_TODOS } from '../actions';
...
         {todos.map((todo, index) => {
           return (
             <div key={index}>
-              <Link to={"/todos/" + todo.id}>{todo.title}</Link>
+              <TodoListItem 
+                todo={todo}
+              />
             </div>
           );
         })}
```
Comment (kenji4569) [inline src/pages/todo/components/TodoList.js:91]: I would like to make the input ui as in gmail todo. But in this pull request, it's ok just to put the add button.

File: src/pages/todo/components/TodoListItem.js (hunk 6 above)
Comment (vinhkq) [inline src/pages/todo/components/TodoListItem.js:29]: Dear Mr @kenji4569 :
Sorry, in this case, i intend use for update, so it not complete yet.

File: src/common/api.js [hunk 8]
```diff
@@ -24,18 +24,20 @@ export function callApi(path) {
     });
 }
 
-export function* fetchApi(requestTypes, path, action) {
+export function* fetchApi(requestTypes, path, method, action) {
   try {
```
Comment (kenji4569) [inline src/common/api.js:28]: @vinhkq remove `console.log`

File: src/pages/todo/components/TodoList.js [hunk 9]
```diff
@@ -13,18 +14,84 @@ import { LIST_TODOS } from '../actions';
...
+      form: { 
+        title: '' 
+      }
+    }
+  }
+
+  componentDidUpdate() {
+    this.refs.textbox.focus()
+  }
+
```
Comment (kenji4569) [inline src/pages/todo/components/TodoList.js:41]: I prefer : `_validateData` -> `_validate`

File: src/pages/todo/components/TodoList.js [hunk 10]
```diff
@@ -13,18 +14,84 @@ import { LIST_TODOS } from '../actions';
   };
 })
 class TodoList extends React.Component {
+  constructor(props) {
+    super(props);
+    this.state = {
+      isEditing: false,
```
Comment (kenji4569) [inline src/pages/todo/components/TodoList.js]: Sorry, I rather prefer `isEditing` -> `editing`

File: src/pages/todo/components/TodoList.js [hunk 11]
```diff
@@ -13,18 +14,84 @@ import { LIST_TODOS } from '../actions';
...
+      return false
+    }
+    return true
+  }
+
+  _submitForm() {
+    // Handle submit
+    if(this._validateData(this.state.form)) {
+      this.props.dispatch({type: CREATE_TODO.REQUEST, data: this.state.form})
+    }
```
Comment (kenji4569) [inline src/pages/todo/components/TodoList.js:53]: Ah, this code does not take care of "async requests": you can't handle errors nor show loader.

To do this, you should first add the following method in this TodoList:

```
componentWillReceiveProps(nextProps) {
  const {error, fetching} = nextProps[CREATE_TODO.BASE];
  if (!error && !fetching && this.state.editing) {
    this.setState({form: {'title': ''}})
    this.setState({editing: false})
  }
  // TODO error handling and show loader (for other issues)
}
```

Off course, remove `this.setState({form: {'title': ''}})` and `this.setState({isEditing: false})` from `_submitForm`.

Then, modify `reduceApi` in common/api.js as follows:

```
export function reduceApi(state, action, requestTypes, onSuccess) {
  switch (action.type) {
    case requestTypes.REQUEST:
      let nextState = Object.assign({}, state);
      nextState[requestTypes.BASE] = {error: null, fetching: true};
      return nextState;
    case requestTypes.FAILURE:
      let nextState = Object.assign({}, state);
      nextState[requestTypes.BASE] = {error: action.error, fetching: false};
      return nextState;
    case requestTypes.SUCCESS:
      let nextState = Object.assign({}, state, onSuccess(action.data));
      nextState[requestTypes.BASE] = {error: null, fetching: false},;
      return nextState;
  }
  return state;
}
```

File: src/pages/todo/components/TodoList.js [hunk 12]
```diff
@@ -2,29 +2,105 @@ import React from 'react';
...
+import { LIST_TODOS, CREATE_TODO } from '../actions';
+import TodoListItem from './TodoListItem';
 
 @provideHooks({
   fetch: ({ dispatch, params: { id } }) => dispatch({type: LIST_TODOS.REQUEST})
 })
 @connect((state) => {
   return {
-    todos: state.todoReducer && state.todoReducer.todos || []
+    todos: state.todoReducer && state.todoReducer.todos || [],
```
Comment (kenji4569) [inline src/pages/todo/components/TodoList.js:13]: Please add fetchState to  todoReducer as follows

```
CREATE_TODO: state.todoReducer || null
```

to

```
fetchState: state.todoReducer || state.todoReducer.fetchState || {}
```

and 

```
if(nextProps[CREATE_TODO.BASE]) {
    const {error, fetching} = nextProps[CREATE_TODO.BASE];
```

to

```
if(nextProps.fetchState[CREATE_TODO.BASE]) {
      const {error, fetching} = nextProps.fetchState[CREATE_TODO.BASE];
```

and 

```
nextState = Object.assign({}, state);
     nextState[requestTypes.BASE] = {error: null, fetching: true};
     return nextState;
```

to

```
let fetchState = {};
fetchState[requestTypes.BASE] = {error: null, fetching: true};
return Object.assign({}, state, fetchState);
```

(same for others)

File: src/common/api.js [hunk 13]
```diff
@@ -24,28 +24,33 @@ export function callApi(path) {
...
   }
 }
 
-export function* watchApi(requestTypes, path) {
-  yield* takeLatest(requestTypes.REQUEST, fetchApi, requestTypes, path)
+export function* watchApi(requestTypes, path, method = 'get') {
+  yield* takeLatest(requestTypes.REQUEST, fetchApi, requestTypes, path, method)
 }
 
 export function reduceApi(state, action, requestTypes, onSuccess) {
```
Comment (kenji4569) [inline src/common/api.js:42]: define `let fetchState = {};` for each case block( for a little performance gain).

**Pull request merged**